projects/uk.ac.man.cs.llvm/src/uk/ac/man/cs/llvm/ir/module/Metadata.java

```diff
@@ -29,28 +29,476 @@
+
+        idx++;
+    }
+
+    protected void createString(long[] args) {
+        String s = "";
+        for (long lc : args) {
+            s += (char) (lc); // TODO: unicode characters?
+        }
+
+        MetadataString node = new MetadataString(s);
+
+        metadata.add(node);
+    }
+
+    protected void createValue(long[] args) {
+        Type t = MetadataArgumentParser.typeValToType(types, symbols, args[0], args[1]);
+
+        MetadataValue node = new MetadataValue(t);
+
+        metadata.add(node);
+    }
+
+    protected void createNode(long[] args) {
+        MetadataNode node = new MetadataNode();
+
+        for (long arg : args) {
+            node.add(metadata.getReference((int) arg));
+        }
+
+        metadata.add(node);
+    }
+
+    protected void createName(long[] args) {
+        String name = "";
+        for (long lc : args) {
+            name += (char) (lc); // TODO: unicode characters?
+        }
+
+        MetadataName node = new MetadataName(name);
+
+        metadata.add(node);
+    }
+
+    protected void createDistinctNode(long[] args) {
+        // [n x md num]
+
+        System.out.println("!" + idx + " - " + MetadataRecord.DISTINCT_NODE + " - " + Arrays.toString(args));
+    }
+
+    protected void createKind(long[] args) {
+        String name = "";
+        for (int i = 1; i < args.length; i++) {
+            name += (char) (args[i]); // TODO: unicode characters?
+        }
+
+        MetadataKind node = new MetadataKind(args[0], name);
+
+        metadata.add(node);
+    }
+
+    protected void createLocation(long[] args) {
+        // int i = 0;
+        //
+        // long distinct = args[argNumber++];
+        // long line = args[i++];
+        // long column = args[i++];
+        // long scope = args[i++];
+        // long inlineAt = args[i++];
+
+        System.out.println("!" + idx + " - " + MetadataRecord.LOCATION + " - " + Arrays.toString(args));
+    }
+
+    protected void createNamedNode(long[] args) {
+        MetadataNamedNode node = new MetadataNamedNode();
+
+        for (long arg : args) {
+            node.add(metadata.getReference((int) arg));
+        }
+
+        metadata.add(node);
+    }
+
+    protected void createAttachment(long[] args) {
+        // [n x mdnodes]
+
+        System.out.println("!" + idx + " - " + MetadataRecord.ATTACHMENT + " - " + Arrays.toString(args));
+    }
+
+    protected void createGenericDebug(long[] args) {
+        // int i = 0;
+        //
+        // long distinct = args[i++];
+        // long tag = args[i++];
+        // long vers = args[i++];
+        // long header = args[i++];
+        // TODO: args[4] // n x md num
+
+        System.out.println("!" + idx + " - " + MetadataRecord.GENERIC_DEBUG + " - " + Arrays.toString(args));
+    }
+
+    protected void createSubrange(long[] args) {
+        MetadataSubrange node = new MetadataSubrange();
+
+        int i = 0;
```

Comment:
I think having named constants for these cases would make the code easier to read and less error-prone (e.g., since you could forgot to increment `i` or mistakenly move a line during refactoring).